presto-pinot/presto-pinot-lib/src/main/java/com/facebook/presto/pinot/PinotClusterInfoFetcher.java

```diff
@@ -0,0 +1,394 @@
+import java.util.concurrent.ThreadLocalRandom;
+import java.util.concurrent.TimeUnit;
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+import java.util.stream.Collectors;
+
+import static com.facebook.airlift.http.client.StringResponseHandler.createStringResponseHandler;
+import static com.google.common.base.MoreObjects.toStringHelper;
+import static java.lang.String.format;
+import static java.util.Objects.requireNonNull;
+import static org.apache.pinot.common.config.TableNameBuilder.extractRawTableName;
+
+public class PinotClusterInfoFetcher
+{
+    private static final Logger log = Logger.get(PinotClusterInfoFetcher.class);
+    private static final String APPLICATION_JSON = "application/json";
+    private static final Pattern BROKER_PATTERN = Pattern.compile("Broker_(.*)_(\\d+)");
+
+    private static final String GET_ALL_TABLES_API_TEMPLATE = "tables";
+    private static final String TABLE_INSTANCES_API_TEMPLATE = "tables/%s/instances";
+    private static final String TABLE_SCHEMA_API_TEMPLATE = "tables/%s/schema";
+    private static final String ROUTING_TABLE_API_TEMPLATE = "debug/routingTable/%s";
+    private static final String TIME_BOUNDARY_API_TEMPLATE = "debug/timeBoundary/%s";
+
+    private final PinotConfig pinotConfig;
+    private final PinotMetrics pinotMetrics;
+    private final HttpClient httpClient;
+
+    private final Ticker ticker = Ticker.systemTicker();
+
+    private final LoadingCache<String, List<String>> brokersForTableCache;
+
+    private final JsonCodec<GetTables> tablesJsonCodec;
+    private final JsonCodec<BrokersForTable> brokersForTableJsonCodec;
+    private final JsonCodec<RoutingTables> routingTablesJsonCodec;
+    private final JsonCodec<TimeBoundary> timeBoundaryJsonCodec;
+
+    @Inject
+    public PinotClusterInfoFetcher(PinotConfig pinotConfig, PinotMetrics pinotMetrics, @ForPinot HttpClient httpClient, JsonCodec<GetTables> tablesJsonCodec, JsonCodec<BrokersForTable> brokersForTableJsonCodec, JsonCodec<RoutingTables> routingTablesJsonCodec, JsonCodec<TimeBoundary> timeBoundaryJsonCodec)
+    {
+        this.brokersForTableJsonCodec = brokersForTableJsonCodec;
+        this.routingTablesJsonCodec = routingTablesJsonCodec;
+        this.timeBoundaryJsonCodec = timeBoundaryJsonCodec;
+        final long cacheExpiryMs = pinotConfig.getMetadataCacheExpiry().roundTo(TimeUnit.MILLISECONDS);
+        this.tablesJsonCodec = requireNonNull(tablesJsonCodec, "json codec");
+
+        this.pinotConfig = requireNonNull(pinotConfig, "pinotConfig");
+        this.pinotMetrics = requireNonNull(pinotMetrics, "pinotMetrics");
+        this.httpClient = requireNonNull(httpClient, "httpClient");
+        this.brokersForTableCache = CacheBuilder.newBuilder().expireAfterWrite(cacheExpiryMs, TimeUnit.MILLISECONDS).build((CacheLoader.from(this::getAllBrokersForTable)));
+    }
+
+    public static JsonCodecBinder addJsonBinders(JsonCodecBinder jsonCodecBinder)
+    {
```

Comment:
You probably don't have to explicit bind the codec here given they are very simple.